Lib/Protocols/IdSSLOpenSSLLoader.pas

```diff
@@ -0,0 +1,372 @@
+        if Result <> NilHandle then
+          break;
+      end;
+    finally
+      LibVersionsList.Free;
+    end;
+  end;
+end;
+
+
+function TOpenSSLLoader.Load: Boolean;
+type
+  TOpenSSL_version_num = function: TIdC_ULONG; cdecl;
+
+var i: integer;
+    OpenSSL_version_num: TOpenSSL_version_num;
+    SSLVersionNo: TIdC_ULONG;
+    LegacyDLLsLoaded: boolean;
+
+  {$IFDEF WINDOWS}
+  procedure ProcessFailedList;
+  var AllowedList: TStringList;
+      i,j: integer;
+  begin
+    AllowedList := TStringList.Create;
+    try
+      AllowedList.Delimiter := ',';
+      AllowedList.StrictDelimiter := true;
+      AllowedList.DelimitedText := LegacyAllowFailed;
+      for i := 0 to AllowedList.Count - 1 do
+      begin
+        j := FFailed.IndexOf(AllowedList[i]);
+        if j <> -1 then
+          FFailed.Delete(j);
+      end;
+    finally
+      AllowedList.Free;
+    end;
+  end;
+  {$ENDIF}
+
+begin                                  //FI:C101
+  Result := not FFailedToLoad;
+  LegacyDLLsLoaded := false;
+  if not Result then
+    Exit;
+  FLibraryLoaded.Lock();
+  try
+    if not FLibraryLoaded.Value then
+    begin
+      FLibCrypto := FindLibrary(CLibCryptoBase + LibSuffix,FSSLLibVersions);
+      FLibSSL := FindLibrary(CLibSSLBase + LibSuffix,FSSLLibVersions);
+      Result := not (FLibCrypto = IdNilHandle) and not (FLibSSL = IdNilHandle);
+      {$IFDEF WINDOWS}
+      if not Result then
+      begin
+        {try the legacy dll names}
+        FLibCrypto := FindLibrary(LegacyLibCrypto,'');
+        FLibSSL := FindLibrary(LegacyLibssl,'');
+        Result := not (FLibCrypto = IdNilHandle) and not (FLibSSL = IdNilHandle);
+        LegacyDLLsLoaded := Result;
+      end;
+      {$ENDIF}
+      if not Result then
+        Exit;
+
+      {Load Version number}
+      OpenSSL_version_num := LoadLibFunction(FLibCrypto, 'OpenSSL_version_num');
+      if not assigned(OpenSSL_version_num) then
+          OpenSSL_version_num := LoadLibFunction(FLibCrypto, 'SSLeay');
+      if not assigned(OpenSSL_version_num) then
+        raise EIdOpenSSLError.Create(ROSSLCantGetSSLVersionNo);
+
+      SSLVersionNo := OpenSSL_version_num();
+      if SSLVersionNo < min_supported_ssl_version then
+        raise EIdOpenSSLError.CreateFmt(RSOSSUnsupportedVersion,[SSLVersionNo]);
+
+      SSLVersionNo := SSLVersionNo shr 12;
+
+
+      for i := 0 to GLibCryptoLoadList.Count - 1 do
+        TOpenSSLLoadProc(GLibCryptoLoadList[i])(FLibCrypto,SSLVersionNo,FFailed);
+
+      for i := 0 to GLibSSLLoadList.Count - 1 do
+         TOpenSSLLoadProc(GLibSSLLoadList[i])(FLibSSL,SSLVersionNo,FFailed);
+
+      {$IFNDEF OPENSSL_IGNORE_MISSING_FUNCTIONS}
+      {$IFDEF WINDOWS}
+      if LegacyDLLsLoaded and (FFailed.Count > 0) then
+        ProcessFailedList;
+      {$ENDIF}
+      FFailedToLoad :=  FFailed.Count > 0;
+      Result := not FFailedToLoad;
+
+      {$IFDEF DEBUG}
+      if not Result then
+      begin
+        writeln;
```

Comment:
This gives me E/A error 105 if I'm running my program as a GUI application.
You might want to use ``OutputDebugMessage`` ($IFDEF Windows) or check for ``{$IFDEF  CONSOLE}``.

Reply:
Point taken - all my testing was on console mode apps or Linux...